src/NetSapiensSharp/Connector.cs

```diff
@@ -82,16 +96,35 @@ public Request<Tresponse> CreateRequest<Tresponse>(string action_path) where Tre
         public IRestResponse<Tresponse> Send<Tresponse>(Request<Tresponse> request)
             where Tresponse : class, new()
         {
+            if (IsAccessTokenExpired())
+            {
+                _SessionToken = null;
+            }
+            int myRetryCount = 1;
+            RequestRetry:
+
             Authenticate();
             request._Request.AddHeader("Authorization", "Bearer " + _SessionToken);
             var r = request._Client.Execute<Tresponse>(request._Request);
-            if (r.StatusCode == System.Net.HttpStatusCode.OK && r.Data == null && typeof(Tresponse) == typeof(Common.OK))
+            if (r.StatusCode == System.Net.HttpStatusCode.Unauthorized && myRetryCount <= _UnauthorizedRetryLimit)
+            {
+                _SessionToken = null;
+                var myOldAuthHeader = request._Request.Parameters.Where(p => p.Name.Equals("Authorization")).First();
+                request._Request.Parameters.Remove(myOldAuthHeader);
+                myRetryCount++;
+                goto RequestRetry;
```

Comment:
Instead of a `goto` (generally concerned a scarier alternative to most other logical structures), consider a `do/while` loop which will allow the code to execute once and run again if it needs to:

```c#
do 
{
    Authenticate();
    request._Request.AddOrUpdateParameter("Authorization", "Bearer " + _SessionToken, ParameterType.HttpHeader);
    var r = request._Client.Execute<Tresponse>(request._Request);

    if (r.StatusCode == System.Net.HttpStatusCode.Unauthorized)
    {
        _SessionToken = null;
        myRetryCount++;
    }
}
while (_SessionToken == null && myRetryCount <= _UnauthorizedRetryLimit);
```

This allows you to execute the authentication code in a loop until we get an authorized token or we run out of retry attempts.